### PR TITLE
[SPARK-12539][follow-up] always sort in partitioning writer

### PR DESCRIPTION
**Pull request by cloud-fan**
address comments in #10498 , especially https://github.com/apache/spark/pull/10498#discussion_r49021259

### CODE REVIEW

**Comment by cloud-fan** [comment]
cc @davies @rxin

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/WriterContainer.scala [hunk 1]
```diff
@@ -461,54 +411,46 @@ private[sql] class DynamicPartitionWriterContainer(
...
+        sorter.insertKV(getSortingKey(currentRow), getOutputRow(currentRow))
+      }
+
+      logInfo(s"Sorting complete. Writing out partition files one at a time.")
+
+      val needNewWriter: (UnsafeRow, UnsafeRow) => Boolean = if (sortColumns.isEmpty) {
+        (key1, key2) => key1 != key2
       } else {
-        null
+        (key1, key2) => key1 == null || !sameBucket(key1, key2)
```
Comment (davies) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/WriterContainer.scala]: Could we generate a partition key (excluding the sorting key)? we just need to compare the key

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/WriterContainer.scala [hunk 2]
```diff
@@ -461,54 +396,55 @@ private[sql] class DynamicPartitionWriterContainer(
...
+        // Given 2 sorting keys, we should ignore the sorting columns to check if these 2 keys
+        // belong to same bucket.
+        val bucketingKey = sortingExpressions.dropRight(sortColumns.length).zipWithIndex.map {
+          case (expr, ordinal) => BoundReference(ordinal, expr.dataType, expr.nullable)
+        }
+        // Use 2 projection here as `UnsafeProjection` will share one result row during execution.
+        val getBucketingKey1 = UnsafeProjection.create(bucketingKey)
+        val getBucketingKey2 = UnsafeProjection.create(bucketingKey)
+
+        (key1, key2) => key1 == null || getBucketingKey1(key1) != getBucketingKey2(key2)
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/WriterContainer.scala]: Is it faster than the previous `sameBucket` function? cc @davies

Comment (davies) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/WriterContainer.scala]: You should save the generated key:

```
currentKey = getBucketingKey(sortedIterator.getKey).copy()
```
compare with: 
```
if (currentKey != getBucketingKey(sortedIterator.getKey)) {
}
```

**Comment by davies** [comment]
LGTM.

**Comment by rxin** [comment]
I manually resolved the conflict during merging and merged it. Thanks.

**Comment by davies** [comment]
@cloud-fan Should we create a patch for 1.6 branch?

**Comment by cloud-fan** [comment]
only backport it without bucket-write?

**Comment by davies** [comment]
Yes, cc @marmbrus

**Comment by rxin** [comment]
We discussed this and decided this is a simple config change users can apply. Changing the behavior automatically for users in a point release is also bad.

**Pull request closed without merging**